### ISSUE TITLE
Clickstream example and WindowStart in the CLICK_USER_SESSIONS

### ISSUE THREAD

**Issue by asaushkin** [confluentinc/examples#462]
I have seen clickstream example and noticed that field `EVENT_TS` in the table `CLICK_USER_SESSIONS` related to creation of the table time but not starting of windows.

I created the CLICK_USER_SESSIONS table at 10:57

```
CREATE TABLE CLICK_USER_SESSIONS AS
    SELECT username, WindowStart() AS EVENT_TS,
        count(*) AS events
        FROM USER_CLICKSTREAM window SESSION (300 second) GROUP BY username;
```

Now at 11:20 I check records in this table:

```
ksql> print CLICK_USER_SESSIONS limit 5;
Format:JSON
{"ROWTIME":1584087650626,"ROWKEY":"DimitriSchenz88\u0000\u0000\u0001p??QB\u0000\u0000\u0001p??-U","USERNAME":"DimitriSchenz88","EVENT_TS":1584086265173,"EVENTS":1358}
{"ROWTIME":1584087650991,"ROWKEY":"Roberto_123\u0000\u0000\u0001p??R?\u0000\u0000\u0001p??**","USERNAME":"Roberto_123","EVENT_TS":1584086264362,"EVENTS":3442}
{"ROWTIME":1584087651103,"ROWKEY":"BenSteins_235\u0000\u0000\u0001p??S\u001F\u0000\u0000\u0001p??/?","USERNAME":"BenSteins_235","EVENT_TS":1584086265760,"EVENTS":1373}
{"ROWTIME":1584087651700,"ROWKEY":"Oriana_70\u0000\u0000\u0001p??Ut\u0000\u0000\u0001p??,\u0007","USERNAME":"Oriana_70","EVENT_TS":1584086264839,"EVENTS":3297}
{"ROWTIME":1584087651901,"ROWKEY":"AndySims_345324\u0000\u0000\u0001p??V=\u0000\u0000\u0001p??*?","USERNAME":"AndySims_345324","EVENT_TS":1584086264535,"EVENTS":1379}
```

Timestamp in these records points to `Fri Mar 13 2020 10:57:44`

What's going wrong? Why the records in the `CLICK_USER_SESSIONS` is not related to starting of the window?

```
11:08 $ cat docker-compose.yml | grep image
    image: confluentinc/cp-zookeeper:5.4.1
    image: confluentinc/cp-server:5.4.1
    image: confluentinc/cp-schema-registry:5.4.1
    image: confluentinc/cp-server-connect:5.4.1
    image: confluentinc/cp-ksql-server:5.4.1
    image: docker.elastic.co/elasticsearch/elasticsearch:6.3.0
    image: grafana/grafana:5.2.4
    image: docker.elastic.co/kibana/kibana:6.3.0
```
![image](https://user-images.githubusercontent.com/471575/76603079-5d2e6500-651d-11ea-93b3-ac244fa8fa56.png)

![image](https://user-images.githubusercontent.com/471575/76603175-8c44d680-651d-11ea-8f04-652ed359310a.png)

**Comment by bbejeck** [comment]
Hi @asaushkin,

Thanks for the question.  By looking at the data you posted, the records do align with the start window time.  To fully explain, let's take a look at a couple of factors at play.


1. The data generation for the click-stream demo continually produces click-stream event records at a rate of one record every 100MS.
2. The table you defined is using `Session` windows with an inactivity gap of 300 seconds (5 minutes).

When you executed the query creating the table, records were immediately available, and the timestamp of the first record is the same as the WindowStart timestamp. So at the rate of data generation, there will never be a gap long enough to qualify for starting a new session, every record coming in will extend the current session, you end up with one big session for each user.

If you were to inspect the `ROWTIME` values from your print statement, you'd see the timestamp aligns with the time you executed the print statement, but the `EVENT_TS` field hasn't changed because the current session is still open.

As an example, I ran the click-stream demo and created the same table from your posted KSQL statement. Then I ran some sample queries with just one user name:

`select * from CLICK_USER_SESSIONS WHERE USERNAME='DimitriSchenz88' and ROWTIME > 1584646128206 EMIT CHANGES LIMIT 20;`

I won't show all the results here, but if you look at the first result:

```
ROWTIME       |                          ROWKEY                                 |     USERNAME      |    EVENT_TS   |  EVENTS 
  
1584647107383 | DimitriSchenz88 : Window{start=1584642466110 end=1584647107383} |  DimitriSchenz88  | 1584642466110 |  10055     

.....

1584647207551 | DimitriSchenz88 : Window{start=1584642466110 end=1584647207551} |  DimitriSchenz88  | 1584642466110 |  10172                          

```

What you'll notice here is the `EVENT_TS` always matches Window start value, but also note that the ROWTIME (timestamp of the record as it arrives) matches the Window end value.  Again since each new record comes within the inactivity gap, it extends the end window time of the session on each arrival. 

HTH,
Bill

**Comment by asaushkin** [comment]
Hi @bbejeck 

Thank you for such a detailed answer!

Your answer is really helpful and I understand session windows much better now.

But, am I right that with that table definition: https://github.com/confluentinc/examples/blob/5.4.1-post/clickstream/ksql/ksql-clickstream-demo/demo/clickstream-schema.sql#L66

I'll never get a diagram like this:
![image](https://user-images.githubusercontent.com/471575/77199753-65c50380-6afa-11ea-84f1-e7833a01dbe5.png)
(Note: screenshot was taken on the [main repository page](https://github.com/confluentinc/examples))

Probably it will be better to fix the table definition [here](https://github.com/confluentinc/examples/blob/5.4.1-post/clickstream/ksql/ksql-clickstream-demo/demo/clickstream-schema.sql#L66) to reduce an inactivity gap of the session window or maybe replace session windows to other type of windows?

**Comment by bbejeck** [comment]
@asaushkin, we've been looking into this some more, and it seems you are correct. We'll update the demo once we figure out the necessary fix.  

I'll be sure to reach out to you once it looks like we have a fix.

Thanks again for reporting this!

**Comment by bbejeck** [comment]
Hi, @asaushkin a fix has been pushed for `5.4.1` (https://github.com/confluentinc/examples/pull/484).  If you re-run the clickstream demo, you should see the expected results.

I'm going to close this issue for now.  Please reach out to us if you have any other questions. and thanks again for taking the time to report the issue.
